pkg/controller/datavolume/controller.go

```diff
@@ -236,63 +231,89 @@ func getDataVolumeOp(dv *cdiv1.DataVolume) dataVolumeOp {
-	var ctx dataVolumeSyncResult
+func (r ReconcilerBase) syncCommon(log logr.Logger, req reconcile.Request, cleanup, prepare dataVolumeSyncResultFunc) (*dataVolumeSyncResult, error) {
+	syncRes, syncErr := r.sync(log, req, cleanup, prepare)
+	if err := r.syncUpdateMeta(log, *syncRes); err != nil {
+		syncErr = err
+	}
+	return syncRes, syncErr
+}
 
+func (r ReconcilerBase) sync(log logr.Logger, req reconcile.Request, cleanup, prepare dataVolumeSyncResultFunc) (*dataVolumeSyncResult, error) {
+	syncRes := &dataVolumeSyncResult{}
 	dv, err := r.getDataVolume(req.NamespacedName)
-	if dv == nil {
-		ctx.res = &reconcile.Result{}
-		return ctx, err
+	if dv == nil || err != nil {
+		syncRes.result = &reconcile.Result{}
+		return syncRes, err
 	}
+	syncRes.dv = dv
+	syncRes.dvMutated = dv.DeepCopy()
+
 	if dv.DeletionTimestamp != nil {
 		log.Info("DataVolume marked for deletion, cleaning up")
 		if cleanup != nil {
-			err := cleanup(dv)
-			return ctx, err
+			err := cleanup(syncRes)
+			return syncRes, err
 		}
-		ctx.res = &reconcile.Result{}
-		return ctx, nil
+		syncRes.result = &reconcile.Result{}
+		return syncRes, nil
 	}
 
 	if prepare != nil {
-		if err := prepare(dv); err != nil {
-			return ctx, err
+		if err := prepare(syncRes); err != nil {
+			return syncRes, err
 		}
 	}
 
-	pvc, err := r.getPVC(dv)
+	syncRes.pvc, err = r.getPVC(dv)
 	if err != nil {
-		return ctx, err
+		return syncRes, err
 	}
-	if pvc != nil {
-		res, err := r.garbageCollect(dv, pvc, log)
-		if err != nil {
-			return ctx, err
+	if syncRes.pvc != nil {
+		if err := r.garbageCollect(syncRes, log); err != nil {
+			return syncRes, err
 		}
-		if res != nil {
-			ctx.res = res
-			return ctx, nil
+		if syncRes.result != nil || syncRes.dv == nil {
+			return syncRes, nil
 		}
-		if err := r.validatePVC(dv, pvc); err != nil {
-			return ctx, err
+		if err := r.validatePVC(dv, syncRes.pvc); err != nil {
+			return syncRes, err
 		}
+		r.handlePrePopulation(syncRes.dvMutated, syncRes.pvc)
 	}
 
-	pvcSpec, err := renderPvcSpec(r.client, r.recorder, log, dv)
+	syncRes.pvcSpec, err = renderPvcSpec(r.client, r.recorder, log, dv)
 	if err != nil {
-		return ctx, err
+		return syncRes, err
+	}
+	return syncRes, nil
+}
+
+func (r ReconcilerBase) syncUpdateMeta(log logr.Logger, syncRes dataVolumeSyncResult) error {
+	if syncRes.dv == nil || syncRes.dvMutated == nil {
+		return nil
+	}
+	if !reflect.DeepEqual(syncRes.dv.Status, syncRes.dvMutated.Status) {
+		return fmt.Errorf("status update is not allowed in sync phase")
+	}
+	if !reflect.DeepEqual(syncRes.dv.ObjectMeta, syncRes.dvMutated.ObjectMeta) {
+		if err := r.updateDataVolume(syncRes.dvMutated); err != nil {
+			r.log.Error(err, "Unable to sync update dv meta", "name", syncRes.dvMutated.Name)
+			return err
+		}
+		// Needed for emitEvent() DeepEqual check
+		objMeta := syncRes.dvMutated.ObjectMeta.DeepCopy()
+		syncRes.dv.ObjectMeta = *objMeta
```

Comment:
I think the right thing to do here is actually:
```
syncRes.dv = syncRes.dvMutated.DeepCopy()
```

Reply:
Problem is what if the spec and metadata is updated?  You have to either specifically disallow that or change the name of this to `syncUpdate`.  Spec updates should not be out of bounds for the controller

Reply:
fixed

Reply:
What do you think about calling this `syncUpdate()` since it's not just about meta

Reply:
@alromeros, ^^^